src/Functions/FunctionStringCompare.cpp

```diff
@@ -0,0 +1,414 @@
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/FunctionsComparison.h>
+#include <Functions/IFunction.h>
+#include <Interpreters/Context.h>
+
+#include <algorithm>
+#include <limits>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+// usage:
+// stringCompare(s, t)
+//  Compare s with t.
+// stringCompare(s, t, offset_s, offset_t, n)
+//  Compare s from offset_s to offset_s + n with t from offset_t to offset_t + n.
+class FunctionStringCompare : public IFunction
+{
+public:
+    static constexpr auto name = "stringCompare";
+    static FunctionPtr create(const ContextPtr /*context*/) { return std::make_shared<FunctionStringCompare>(); }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 0; }
+    bool isVariadic() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        if (arguments.size() != 2 && arguments.size() != 5)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires 2 or 5 arguments", getName());
+
+        FunctionArgumentDescriptors mandatory_args{
+            {"string1", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String"},
+            {"string2", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String"},
+        };
+        FunctionArgumentDescriptors optional_args{
+            {"string1_offset", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt), static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn), "const UInt*"},
+            {"string2_offset", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt), static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn), "const UInt*"},
+            {"num_bytes", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt), static_cast<FunctionArgumentDescriptor::ColumnValidator>(&FunctionStringCompare::isConstColumn), "const UInt*"},
+        };
+        validateFunctionArguments(*this, arguments, mandatory_args, optional_args);
+
+        return std::make_shared<DataTypeInt8>();
+    }
+
+    DataTypePtr getReturnTypeForDefaultImplementationForDynamic() const override { return std::make_shared<DataTypeInt8>(); }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & /*result_type*/, size_t input_rows_count) const override
+    {
+        chassert(arguments[0].column->size() == input_rows_count);
+        auto col_result = ColumnInt8::create();
+        if (!input_rows_count)
```

Comment:
Let's spend an extra five keystrokes for ` == 0` to make more clear that `input_rows_count` is not a boolean variable.